{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyPgB35F6PHD3g/5sHjD6s9Z",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/Pmilivojevic/Bioinformatics/blob/main/bioinformatics_chapter2.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "p9iLBERannYx",
        "outputId": "78a00203-d9f0-49d6-93a8-4013db7cf5d8"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mounted at /content/drive/\n"
          ]
        }
      ],
      "source": [
        "import numpy as np\n",
        "import time\n",
        "import copy\n",
        "import random\n",
        "from google.colab import drive\n",
        "drive.mount('/content/drive/')\n",
        "nucleotides= ['A', 'C', 'G', 'T']"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def symbolToNum(c):\n",
        "  if c == 'A':\n",
        "    return 0\n",
        "  elif c == 'C':\n",
        "    return 1\n",
        "  elif c == 'G':\n",
        "    return 2\n",
        "  else:\n",
        "    return 3\n",
        "def patternToNum(pattern):\n",
        "  if len(pattern) == 0:\n",
        "    return 0\n",
        "  else:\n",
        "    symbol= pattern[len(pattern)-1]\n",
        "    prefix=  pattern[:len(pattern)-1]\n",
        "    return 4*patternToNum(prefix) + symbolToNum(symbol)\n",
        "def numToSymbol(n):\n",
        "  if n == 0:\n",
        "    return 'A'\n",
        "  elif n == 1:\n",
        "    return 'C'\n",
        "  elif n == 2:\n",
        "    return 'G'\n",
        "  else:\n",
        "    return 'T'\n",
        "def numToPattern(index, k):\n",
        "  if k == 1:\n",
        "    return numToSymbol(index)\n",
        "  else:\n",
        "    prefixInd= int(index/4)\n",
        "    remainder= index%4\n",
        "    symbol= numToSymbol(remainder)\n",
        "    return numToPattern(prefixInd, k-1) + symbol\n",
        "def immediateNeighbors(pattern):\n",
        "  neighborhood= [pattern]\n",
        "\n",
        "  for i in range(len(pattern)):\n",
        "    symbol= pattern[i]\n",
        "\n",
        "    for X in nucleotides:\n",
        "      if X != pattern[i]:\n",
        "        sample= pattern[0:i]+X+pattern[i+1:len(pattern)]\n",
        "        neighborhood.append(sample)\n",
        "\n",
        "  return neighborhood\n",
        "def sufix(text):\n",
        "  return text[1:len(text)]\n",
        "def neighbors(pattern, d):\n",
        "  if d == 0:\n",
        "    return pattern\n",
        "\n",
        "  if len(pattern) == 1:\n",
        "    return nucleotides\n",
        "\n",
        "  neighborhood= []\n",
        "  sufixNeighbors= neighbors(sufix(pattern), d)\n",
        "\n",
        "  for s in sufixNeighbors:\n",
        "    if hammingDistance(s, sufix(pattern)) < d:\n",
        "      for X in nucleotides:\n",
        "        neighborhood.append(X+s)\n",
        "    else:\n",
        "      neighborhood.append(pattern[0]+s)\n",
        "\n",
        "  return neighborhood"
      ],
      "metadata": {
        "id": "hSOYONhi_Kxg"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def motifEnumeration(DNA, k, d):\n",
        "  patterns= []\n",
        "\n",
        "  for seq in DNA:\n",
        "    for i in range(len(seq)-k+1):\n",
        "      for pattern in neighbors(seq[i:i+k], d):\n",
        "        for string in DNA:\n",
        "          exist= False\n",
        "\n",
        "          for j in range(len(string)-k+1):\n",
        "            if hammingDistance(pattern, string[j:j+k]) <= d:\n",
        "              exist= True\n",
        "              break\n",
        "          if not exist:\n",
        "            break\n",
        "\n",
        "        if exist:\n",
        "         # print(pattern)\n",
        "          ex= False\n",
        "\n",
        "          for ptt in patterns:\n",
        "            if pattern == ptt:\n",
        "              ex= True\n",
        "              break\n",
        "          if not ex:\n",
        "            patterns.append(pattern)\n",
        "\n",
        "  return patterns"
      ],
      "metadata": {
        "id": "CIIX4gJwW3-b"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def hammingDistance(p, q):\n",
        "  d= 0\n",
        "  for i in range(len(p)):\n",
        "    if p[i] != q[i]:\n",
        "      d+= 1\n",
        "\n",
        "  return d"
      ],
      "metadata": {
        "id": "Fj_w7y5en1EM"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def d(pattern, seq):\n",
        "  minDist= hammingDistance(pattern, seq[0:len(pattern)])\n",
        "\n",
        "  for i in range(1, len(seq)-len(pattern)+1):\n",
        "    if hammingDistance(pattern, seq[i:i+len(pattern)]) < minDist:\n",
        "      minDist= hammingDistance(pattern, seq[i:i+len(pattern)])\n",
        "\n",
        "  return minDist\n",
        "\n",
        "def dDNA(pattern, DNA):\n",
        "  minDistSum= 0\n",
        "\n",
        "  for dna in DNA:\n",
        "    minDistSum+= d(pattern, dna)\n",
        "\n",
        "  return minDistSum"
      ],
      "metadata": {
        "id": "rGA-MBDZoC0U"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "seq= 'TGAACTTTACCTCGATGAGGCCTTGTCGGTCTTCGAGGACTTGCTCGGCCGGCCCCCGCAGCGCTGCTTTCGTGAGTGTCAAAGCTGTTGAGACGGAGGTTGAAT AGGTGTTCTGGCCCACCAACCGATCTATATGTGGGCCTGTCGGCACCGATGGTCCTTACCCTGTCCTGTAGGCAATTTGACATAAAGTCGACGATTGGCCACCAG CTAGGGGTATTGAAACAAAGCGTACTCATTTTGATGGCCCTCAAAGGAATTATGTGACCCGCAGCTAATAGCATTGGGGGAAAGTACTTATTGGGCTGCGTGAGG TGGAAATCTCTTTCTAAGCTCGTTCTTCTCATATACGGATATTCAAGGGACACGGTTGTCGCGTTATCCAATGTATCTCGGTACGAGTGCCGTGCGCCCCATGCC TAGGGGCAATGCTCCCTGAACTGGCTAACAAGTTTGCGCCCTAGCGGACGCCAAGCAACCATGCGGATGTCTAGAGAGACTTTAGTTGCCTCAGACGTTTTAGTA CTGCCTTCCTCATTCCAAGATTGGTAACACACTGGCCAATAACAGGAATGGTAGCATACAGCGCACTATCTATACACCTTCCTACGTCTTACCAGATAGGTTTAA CAGTCTATTCATTTTACCCTCAATCGCATTGGGTTGATCATGGGGATACGCGTACGCCACCGCTACAGCATTGTCCCATGCTTTCAATGAGGGTCAGCTAACCAA CAACCTCTTAAACGAGGAACCATGGGACCGTGTCTGCGGCTCTGCCACAAACTATGCGAAGTTCGAGGCGATTCCTCCCGTCATGCATCCTAGCTATGGCCTGGT CTATTCGATGTGTCACTTTCTTTGCTCCAGTCTGTTTCGCACCGATCTTTTTCAACGACTACATGCACGTTTTGTGTGCTTCGTACTAAGCACCGACTTATAGTT TTTACTTTCCGGACGACTCAGATTCCTAGACAGCTCATACGATGCTAGATCTTTGTGATCCAATAAATGAGCGGACTGCATCTAGGGCAAGCTATTACATGATTT TTGCGGCATACATTGCGACAGGAGACCCTAGTGCCCAAACTAATAACGGTACTACCGCATGCCTACTCACTCCGATATCATGCTTCATCAAGCGTCTATCACGGG TGCGTGGTAAGCGATCCTAAGTATTTCAATGATCGTAAAGGGTGTTTTATTGCTGGGAAAAACTGACACTATAGTAGATTGAACCCGGGGGGGATTCACCGTACA ACGAAGCGTCTTAATGCGGGAAGTGGGGGCAACAACGAGGGTACTGAAATCTAAAGTTCCTATGATATCGGTATTAGCACAGTTTACGATCATCGCGCGGGTTTC ACGCGAACACTAAAAGGGCCTCGCGCGTTCCGCAGCTGCTACGAGGCAGCTCGCCGAGGTCACGCCATTCCATTCTTCAGTTAATTGTTCCTCCCCTATTCTTCA TAGCGGGTCTTACCGTCGAAATTTCTATCCGGGCCGAAGTTTAGGCAGAGGTATGGTGAGTGACGTGGAGCCATCGCTACAAGATTATCAGTTAGTCCATTCGCA ACTATCAATATCCTTTATCGAAGGTAACCAAGGCAGCAGAACCGACGCGGTCATATACCCCAATTCCCGTGGTATCTTGTAATTGGCAATTTCCCTCTCTCTCCG CAGCGATCGCCATTTTTACAGGCAAAGAGGGCGTATATCCACGATTCACCCCGCTTAATCACCGGAGCTGCTTTCTGTAAGCGGGTACGGTCCTTGTCTTCCCCT AACAGCGCATTAAGCATGATCGACGAAGTTGTGATTGCAAACATAGCACGTCATTTAGAATTCAATCCGTTAGAATTATGTAAGCAGAGGTATTCAAGGTGTCGC GCTGGTAGTAAAAGGGCGAAAAGGCCACAGCAGTCCACCCAGGTGCGTGGAATTGCGACTGGGATAAGAGCTTTAGCCTCGCTGTATGCGACGCTAGCTAAATCC TAAAGTCGCGAAGTCGCAGCTAGGATAGCGACCCGTTGCAGATAGTTTATCCTCCGAGACCTCTGTAAGATACTAATCCCTATCGCTAGTGCTCCGACATTGGCT AGCCTTTGGAAGGACCAGTGTAGACATAGTCCATGTCCACCCAGCCCGAGGCTCCCTGCAGAACCATCGTGGGCCAGGATGAGCCTACGGCGGTCGGTCTGCTAG TCTTATATGCGTGTCAAGGGGTGCACGATTAAGTGATCAGGGTGAATCGGGATACCTGATGATGGTTAAAAGCCAACTCGTCACAAGTTAATGTTAGGGACCAGA GAGACAGTCATCTGTGAGACATACATCGAACCTGATTGAAAGGTGTATGAATTCTTGGTTTGCTCCGTAGAGCAGACGTGCTTAGCTAGCGAATGATTACCTCCC CGTTGAATACCCTCCTTTCTTTACCAACCCGACGAGAAAGCACCATACCCGATGAATCAGGTTGATATGTGTACACGCGTCAAATGAAAACGGGGAATCCCTAAG GTGTGTATCTTGCTACTACTCATGATTCCAACCCTGACCTCTTATCGTATCGCAAGCCCATGCATAGATCTGCGTGAGCGACCTACGCTCCGCCTACTGTTCTAG CCAATGCCATTTAATTGCTGACCTCTACAGCCGGGGGTAAAGGCACGCACGCCCGCACTCTTGTCTCTCCAAGCGTAGGTACGCATACTAAGGATACCTTGTTAC TCCAGCCGTGTCCTGGGCCCGGGCGTAAAAAAAGCAAGATTGTACTTAGTATGGTACATATGCGCGCGAGGCCTACCAAGCCGTGGGGAGCTATCATAGCCCTAG ACATTGAGAAGGGCTTCACCAGATTCGGGATAGGGAAGGAGTAGCGCCGCCCTTTCACCGAGCAAAAAGGGAGAGTATAATTCGCAAGTCCTTGAGCGTGGCTGA TGGCCTAGCTGAATATTATATTGTGTTATCAGAAGCCATACAAGCCCGTGAACAGTGGCTCTCTGAATCCGCGGATCTGAACGGGGGATTTTTGCCGTGTTTAAC TTTCATCTGTATCGGTCGACTATCCAACGCACAGTAATCATTACCGCCATTTGGTCACAGGGCAACGTATCTGATAGATTAGCACAACTAAGCTGGTGTGTCATT GGGTGAGATTTTACGGGGCGTTGAGTCTGAAGATTTAAACGGACTCTATTGCAGTAGAGTCTAGTCAGGGGATGTTTGAATGAGACGATTGGACTAACGCGGTGA CTCAGAATGGCGTTAGCCCCACGTGAGGCTGCGTATAACCGGAGAAGATCGTGCGCGCCGCGATTGTACCGGCTGGTTTTGATTTTGAATAAGTCGTACTAGATG GCATCGGTCAATGGCACCAACAGCCCGGCTATCAGAGACGGAAAGCCTCCAGAACTTGGACGAAGGCCGATTCGGTAATGCACCAAACGGCGTCCGCGCATGGCA AGCAATCTTCGTTCGCAAGGACTACCCGATACTAGGGTCTAGAACTGTAGCATTATCAGGCTCGAGACGCTTAGATCACACAGCTGCTAGTGACCAGCTCTCGTC ACGAACAGGCCCACCTCCCTCTATTACGAGAGAGTCGACCACTGCTTCACTTCTGGCACGCGGGTCGTGCCAGGGGTATTTGAAACATACTACGCATGCACTAAG CTCGGGTTAGGTTCAAGCAGGGACTCGGGGCATAGACCATACCATCGGAGACGTAGATTGGTATATAAGCTACACGAGGGTTAGACGAAAGAACAACAGATCAAA TCGCAATCGTTTCATCCCGATTTACTTTTCCGTATATCAGTGATCCGCATTAGCGACGGGGACGGTGCATCGTGGGTTCCGACTCGTCCAGATGCATGCTATCGG GGACTGGAACAAAGCTTACTGGGTCTGAAGCAGCACTTTTTTCACAACTGTGCCTGGGCTAAGTGTGACACGCCTGGAGGGTGCTTTATATGGTCATTTAGAAGT'\n",
        "DNA= []\n",
        "string= ''\n",
        "for c in seq:\n",
        "  if c != ' ':\n",
        "    string+= c\n",
        "  else:\n",
        "    DNA.append(string)\n",
        "    string= ''\n",
        "DNA.append(string)\n",
        "\n",
        "pattern= 'TTGTTC'\n",
        "\n",
        "print(dDNA(pattern, DNA))"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "BQN8848AUFUq",
        "outputId": "25268547-6fae-49f1-b833-edf7ef3e8d7d"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "62\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def medianString(DNA, k):\n",
        "  patterns= []\n",
        "  for i in range(1,4**k):\n",
        "    patterns.append(numToPattern(i,k))\n",
        "\n",
        "  minDist= dDNA(numToPattern(0,k), DNA)\n",
        "  madian= numToPattern(0,k)\n",
        "\n",
        "  for pattern in patterns:\n",
        "    if dDNA(pattern, DNA) < minDist:\n",
        "      minDist= dDNA(pattern, DNA)\n",
        "      median= pattern\n",
        "\n",
        "  return median"
      ],
      "metadata": {
        "id": "7zNSaOdZqJlH"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def profile(seqs):\n",
        "  if len(seqs[0]) == 1:\n",
        "    matProf= np.ones((4, len(seqs)))\n",
        "    for i in range(len(seqs)):\n",
        "      if seqs[i] == 'A':\n",
        "        matProf[0][i]+= 1\n",
        "      elif seqs[i] == 'C':\n",
        "        matProf[1][i]+= 1\n",
        "      elif seqs[i] == 'G':\n",
        "        matProf[2][i]+= 1\n",
        "      else:\n",
        "        matProf[3][i]+= 1\n",
        "    matProf/= 5\n",
        "  else:\n",
        "    matProf= np.zeros((4, len(seqs[0])))\n",
        "\n",
        "    for i in range(len(seqs[0])):\n",
        "      A= 0\n",
        "      C= 0\n",
        "      G= 0\n",
        "      T= 0\n",
        "      for j in range(len(seqs)):\n",
        "        if seqs[j][i] == 'A':\n",
        "          A+= 1\n",
        "        elif seqs[j][i] == 'C':\n",
        "          C+= 1\n",
        "        elif seqs[j][i] == 'G':\n",
        "          G+= 1\n",
        "        else:\n",
        "          T+= 1\n",
        "\n",
        "      matProf[0][i]= (A+1)/(len(seqs)+4)\n",
        "      matProf[1][i]= (C+1)/(len(seqs)+4)\n",
        "      matProf[2][i]= (G+1)/(len(seqs)+4)\n",
        "      matProf[3][i]= (T+1)/(len(seqs)+4)\n",
        "\n",
        "      '''matProf[0][i]= A/len(seqs)\n",
        "      matProf[1][i]= C/len(seqs)\n",
        "      matProf[2][i]= G/len(seqs)\n",
        "      matProf[3][i]= T/len(seqs)'''\n",
        "\n",
        "  return matProf\n",
        "\n",
        "def consensus(prof):\n",
        "  consensusStr= ''\n",
        "\n",
        "  for i in range(len(prof[0])):\n",
        "    for j in range(4):\n",
        "      if prof[j][i] == max([prof[0][i], prof[1][i], prof[2][i], prof[3][i]]):\n",
        "        if j == 0:\n",
        "          consensusStr+= 'A'\n",
        "          break\n",
        "        elif j == 1:\n",
        "          consensusStr+= 'C'\n",
        "          break\n",
        "        elif j == 2:\n",
        "          consensusStr+= 'G'\n",
        "          break\n",
        "        elif j == 3:\n",
        "          consensusStr+= 'T'\n",
        "          break\n",
        "\n",
        "  return consensusStr\n",
        "\n",
        "def score(motifs, conKmer):\n",
        "  sc= 0\n",
        "  for i in range(len(motifs)):\n",
        "    for j in range(len(motifs[0])):\n",
        "      if motifs[i][j] != conKmer[j]:\n",
        "        sc+= 1\n",
        "\n",
        "  return sc\n",
        "\n",
        "def motifPr(profMat, kmer):\n",
        "  if kmer[0] == 'A':\n",
        "    pr= profMat[0][0]\n",
        "  elif kmer[0] == 'C':\n",
        "    pr= profMat[1][0]\n",
        "  elif kmer[0] == 'G':\n",
        "    pr= profMat[2][0]\n",
        "  else:\n",
        "    pr= profMat[3][0]\n",
        "\n",
        "  for i in range(1, len(kmer)):\n",
        "    if kmer[i] == 'A':\n",
        "      pr*= profMat[0][i]\n",
        "    elif kmer[i] == 'C':\n",
        "      pr*= profMat[1][i]\n",
        "    elif kmer[i] == 'G':\n",
        "      pr*= profMat[2][i]\n",
        "    elif kmer[i] == 'T':\n",
        "      pr*= profMat[3][i]\n",
        "\n",
        "  return pr\n",
        "\n",
        "def mostPrMotif(profMat, seq):\n",
        "  k= len(profMat[0])\n",
        "  maxPr= motifPr(profMat, seq[0:k])\n",
        "  prMotif= seq[0:k]\n",
        "\n",
        "  for i in range(1, len(seq)-k+1):\n",
        "    if motifPr(profMat, seq[i:i+k]) > maxPr:\n",
        "      maxPr= motifPr(profMat, seq[i:i+k])\n",
        "      prMotif= seq[i:i+k]\n",
        "\n",
        "  return prMotif\n",
        "\n",
        "def profRandMotif(profMat, seq):\n",
        "  k= len(profMat[0])\n",
        "  pr= []\n",
        "  kmers= []\n",
        "  for i in range(len(seq)-k+1):\n",
        "    pr.append(motifPr(profMat, seq[i:i+k]))\n",
        "    kmers.append(seq[i:i+k])\n",
        "\n",
        "  pr= pr/sum(pr)\n",
        "\n",
        "  return random.choices(kmers, weights= pr)[0]"
      ],
      "metadata": {
        "id": "SvK9wRVS-sRI"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def makeMotifs(profMat, DNA):\n",
        "  motifs= []\n",
        "\n",
        "  for seq in DNA:\n",
        "    motifs.append(mostPrMotif(profMat, seq))\n",
        "\n",
        "  return motifs\n",
        "\n",
        "def randomizedMotifSearch(DNA, k):\n",
        "  motifs= []\n",
        "  for seq in DNA:\n",
        "    i= random.randint(0, len(seq)-k)\n",
        "    motifs.append(seq[i:i+k])\n",
        "\n",
        "  bestMotifs= motifs\n",
        "\n",
        "  while True:\n",
        "    m= profile(motifs)\n",
        "    motifs= makeMotifs(m, DNA)\n",
        "    s= score(bestMotifs, consensus(m))\n",
        "    if score(motifs, consensus(m)) < s:\n",
        "      bestMotifs= motifs\n",
        "    else:\n",
        "      return bestMotifs, s\n",
        "\n",
        "def gibsSampler(DNA, k, n):\n",
        "  motifs= []\n",
        "  for seq in DNA:\n",
        "    i= random.randint(0, len(seq)-k)\n",
        "    motifs.append(seq[i:i+k])\n",
        "\n",
        "  bestMotifs= motifs\n",
        "\n",
        "  for j in range(n):\n",
        "    l= random.randint(0, len(DNA)-1)\n",
        "\n",
        "    motifsNew= []\n",
        "    for i in range(len(motifs)):\n",
        "      if i != l:\n",
        "        motifsNew.append(motifs[i])\n",
        "\n",
        "    m= profile(motifsNew)\n",
        "    motifs[l]= profRandMotif(m, DNA[l])\n",
        "\n",
        "    if score(motifs, consensus(m)) < score(bestMotifs, consensus(m)):\n",
        "      bestMotifs= motifs\n",
        "\n",
        "  return bestMotifs"
      ],
      "metadata": {
        "id": "tbxwddM2otlm"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def greedyMotifSearch(DNA, k):\n",
        "  bestMotifs= []\n",
        "\n",
        "  for i in range(len(DNA)):\n",
        "    bestMotifs.append(DNA[i][0:k])\n",
        "\n",
        "  motifs= []\n",
        "\n",
        "  for i in range(len(DNA[0])-k+1):\n",
        "    motifs.append(DNA[0][i:i+k])\n",
        "\n",
        "    for i in range(1,len(DNA)):\n",
        "      m= profile(motifs)\n",
        "\n",
        "      motifs.append(mostPrMotif(m, DNA[i]))\n",
        "    print('B', bestMotifs)\n",
        "    print(score(bestMotifs, consensus(profile(bestMotifs))))\n",
        "    print('M', motifs)\n",
        "    print(score(motifs, consensus(profile(motifs))))\n",
        "    if score(motifs, consensus(profile(motifs))) < score(bestMotifs, consensus(profile(bestMotifs))):\n",
        "      bestMotifs= motifs\n",
        "    motifs= []\n",
        "\n",
        "  return bestMotifs"
      ],
      "metadata": {
        "id": "d825ZJ0aimIo"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "dna1= []\n",
        "dna1.append('TTAC')\n",
        "m= profile(dna1)\n",
        "dna= 'GATGTCTGTC'\n",
        "print(m)\n",
        "print('mostPrMotif: ', mostPrMotif(m, dna))\n",
        "print('-------------------')\n",
        "for i in range(10):\n",
        "  print('profRandMotif: ', profRandMotif(m, dna))"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "8Truicgnapz-",
        "outputId": "89bf95d8-bcb6-46f6-d236-86fb78ffdb1f"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[[0.2 0.2 0.4 0.2]\n",
            " [0.2 0.2 0.2 0.4]\n",
            " [0.2 0.2 0.2 0.2]\n",
            " [0.4 0.4 0.2 0.2]]\n",
            "mostPrMotif:  TGTC\n",
            "-------------------\n",
            "profRandMotif:  TCTG\n",
            "profRandMotif:  TGTC\n",
            "profRandMotif:  TCTG\n",
            "profRandMotif:  TGTC\n",
            "profRandMotif:  TCTG\n",
            "profRandMotif:  TGTC\n",
            "profRandMotif:  TGTC\n",
            "profRandMotif:  TCTG\n",
            "profRandMotif:  TGTC\n",
            "profRandMotif:  TGTC\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def Count(Motifs):\n",
        "    count = {} # initializing the count dictionary\n",
        "    k = len(Motifs[0])\n",
        "    for symbol in 'ACGT':\n",
        "        count[symbol] = []\n",
        "        for j in range(k):\n",
        "             count[symbol].append(0)\n",
        "\n",
        "    t = len(Motifs)\n",
        "    for i in range(t):\n",
        "        for j in range(k):\n",
        "            symbol = Motifs[i][j]\n",
        "            count[symbol][j] += 1\n",
        "\n",
        "    return count\n",
        "\n",
        "def Profile(Motifs):\n",
        "    t = len(Motifs)\n",
        "    k = len(Motifs[0])\n",
        "\n",
        "    counts= Count(Motifs)\n",
        "    profile = {'A': list(np.array(counts['A'])/t), 'C': list(np.array(counts['C'])/t), 'G': list(np.array(counts['G'])/t), 'T': list(np.array(counts['T'])/t)}\n",
        "\n",
        "    return profile\n",
        "\n",
        "def Consensus(Motifs):\n",
        "    k = len(Motifs[0])\n",
        "    count = Count(Motifs)\n",
        "    consensus = \"\"\n",
        "    for j in range(k):\n",
        "        m = 0\n",
        "        frequentSymbol = \"\"\n",
        "        for symbol in \"ACGT\":\n",
        "            if count[symbol][j] > m:\n",
        "                m = count[symbol][j]\n",
        "                frequentSymbol = symbol\n",
        "        consensus += frequentSymbol\n",
        "\n",
        "    return consensus\n",
        "\n",
        "def Score(Motifs):\n",
        "    sc= 0\n",
        "    conKmer= Consensus(Motifs)\n",
        "    for i in range(len(Motifs)):\n",
        "        for j in range(len(Motifs[0])):\n",
        "            if Motifs[i][j] != conKmer[j]:\n",
        "                sc+= 1\n",
        "\n",
        "    return sc\n",
        "\n",
        "def Pr(Text, Profile):\n",
        "    pr= 1\n",
        "\n",
        "    for i in range(len(Text)):\n",
        "        if Text[i] == 'A':\n",
        "            pr*= Profile['A'][i]\n",
        "        elif Text[i] == 'C':\n",
        "            pr*= Profile['C'][i]\n",
        "        elif Text[i] == 'G':\n",
        "            pr*= Profile['G'][i]\n",
        "        else:\n",
        "            pr*= Profile['T'][i]\n",
        "    return pr\n",
        "\n",
        "def ProfileMostProbableKmer(text, k, profile):\n",
        "    maxPr= Pr(text[0:k], profile)\n",
        "    prMotif= text[0:k]\n",
        "\n",
        "    for i in range(1, len(text)-k+1):\n",
        "        if Pr(text[i:i+k], profile) > maxPr:\n",
        "            maxPr= Pr(text[i:i+k], profile)\n",
        "            prMotif= text[i:i+k]\n",
        "\n",
        "    return prMotif\n",
        "\n",
        "def GreedyMotifSearch(Dna, k, t):\n",
        "    BestMotifs = [] # output variable\n",
        "    # your code here\n",
        "    for i in range(t):\n",
        "        BestMotifs.append(Dna[i][0:k])\n",
        "\n",
        "    motifs= []\n",
        "\n",
        "    for i in range(len(Dna[0])-k+1):\n",
        "        motifs.append(Dna[0][i:i+k])\n",
        "\n",
        "        for i in range(1,t):\n",
        "            m= Profile(motifs)\n",
        "            motifs.append(ProfileMostProbableKmer(Dna[i], k, m))\n",
        "\n",
        "        if Score(motifs) < Score(BestMotifs):\n",
        "            BestMotifs= motifs\n",
        "        motifs= []\n",
        "\n",
        "    return BestMotifs\n",
        "\n",
        "def CountWithPseudocounts(Motifs):\n",
        "    t = len(Motifs)\n",
        "    k = len(Motifs[0])\n",
        "    # your code here\n",
        "    A= []\n",
        "    C= []\n",
        "    G= []\n",
        "    T= []\n",
        "    for i in range(k):\n",
        "        countA= 1\n",
        "        countC= 1\n",
        "        countG= 1\n",
        "        countT= 1\n",
        "        for j in range(t):\n",
        "            if Motifs[j][i] == 'A':\n",
        "                countA+= 1\n",
        "            elif Motifs[j][i] == 'C':\n",
        "                countC+= 1\n",
        "            elif Motifs[j][i] == 'G':\n",
        "                countG+= 1\n",
        "            else:\n",
        "                countT+= 1\n",
        "        A.append(countA)\n",
        "        C.append(countC)\n",
        "        G.append(countG)\n",
        "        T.append(countT)\n",
        "\n",
        "    count = {'A': A, 'C': C, 'G': G, 'T': T} # output variable\n",
        "    return count\n",
        "\n",
        "def ProfileWithPseudocounts(Motifs):\n",
        "    t = len(Motifs)\n",
        "    k = len(Motifs[0])\n",
        "\n",
        "    counts= CountWithPseudocounts(Motifs)\n",
        "    A= list(np.array(counts['A'])/(t+4))\n",
        "    C= list(np.array(counts['C'])/(t+4))\n",
        "    G= list(np.array(counts['G'])/(t+4))\n",
        "    profile = {'A': A, 'C': C, 'G': G, 'T': list(np.array(counts['T'])/(t+4))} # output variable\n",
        "    return profile\n",
        "\n",
        "def MotifPr(profMat, kmer):\n",
        "    pr= 1\n",
        "\n",
        "    for i in range(len(kmer)):\n",
        "        if kmer[i] == 'A':\n",
        "            pr*= profMat['A'][i]\n",
        "        elif kmer[i] == 'C':\n",
        "            pr*= profMat['C'][i]\n",
        "        elif kmer[i] == 'G':\n",
        "            pr*= profMat['G'][i]\n",
        "        else:\n",
        "            pr*= profMat['T'][i]\n",
        "    return pr\n",
        "\n",
        "def ProfileMostProbableKmer(text, k, profile):\n",
        "    maxPr= Pr(text[0:k], profile)\n",
        "    prMotif= text[0:k]\n",
        "\n",
        "    for i in range(1, len(text)-k+1):\n",
        "        if Pr(text[i:i+k], profile) > maxPr:\n",
        "            maxPr= Pr(text[i:i+k], profile)\n",
        "            prMotif= text[i:i+k]\n",
        "\n",
        "    return prMotif\n",
        "\n",
        "def Motifs(Profile, Dna):\n",
        "    # insert your code here\n",
        "    motifs= []\n",
        "\n",
        "    for seq in Dna:\n",
        "        motifs.append(MostPrMotif(Profile, seq))\n",
        "\n",
        "    return motifs\n",
        "\n",
        "def Consensus(prof):\n",
        "    consensusStr= ''\n",
        "\n",
        "    for i in range(len(prof['A'])):\n",
        "        Max= max([prof['A'][i], prof['C'][i], prof['G'][i], prof['T'][i]])\n",
        "        if prof['A'][i] == Max:\n",
        "            consensusStr+= 'A'\n",
        "        elif prof['C'][i] == Max:\n",
        "            consensusStr+= 'C'\n",
        "        elif prof['G'][i] == Max:\n",
        "            consensusStr+= 'G'\n",
        "        elif prof['T'][i] == Max:\n",
        "            consensusStr+= 'T'\n",
        "\n",
        "    return consensusStr\n",
        "\n",
        "def Score(motifs):\n",
        "    sc= 0\n",
        "    conKmer= Consensus(ProfileWithPseudocounts(motifs))\n",
        "    for i in range(len(motifs)):\n",
        "        for j in range(len(motifs[0])):\n",
        "            if motifs[i][j] != conKmer[j]:\n",
        "                sc+= 1\n",
        "\n",
        "    return sc\n",
        "\n",
        "def GreedyMotifSearchWithPseudocounts(Dna, k, t):\n",
        "    BestMotifs = [] # output variable\n",
        "    # your code here\n",
        "    for i in range(t):\n",
        "        BestMotifs.append(Dna[i][0:k])\n",
        "\n",
        "    motifs= []\n",
        "\n",
        "    for i in range(len(Dna[0])-k+1):\n",
        "        motifs.append(Dna[0][i:i+k])\n",
        "\n",
        "        for i in range(1,t):\n",
        "            m= ProfileWithPseudocounts(motifs)\n",
        "            motifs.append(MostPrMotif(m, Dna[i]))\n",
        "\n",
        "        if Score(motifs) < Score(BestMotifs):\n",
        "            BestMotifs= motifs\n",
        "        motifs= []\n",
        "\n",
        "    return BestMotifs\n",
        "\n",
        "def RandomMotifs(Dna, k, t):\n",
        "    motifs= []\n",
        "\n",
        "    for seq in Dna:\n",
        "        i= random.randint(0, len(seq)-k)\n",
        "        motifs.append(seq[i:i+k])\n",
        "\n",
        "    return motifs\n",
        "\n",
        "def RandomizedMotifSearch(Dna, k, t):\n",
        "    # insert your code here\n",
        "    motifs= RandomMotifs(Dna, k, t)\n",
        "    BestMotifs= motifs\n",
        "\n",
        "    while True:\n",
        "        profile= ProfileWithPseudocounts(motifs)\n",
        "        motifs= Motifs(profile, Dna)\n",
        "        if Score(motifs) < Score(BestMotifs):\n",
        "            BestMotifs= motifs\n",
        "        else:\n",
        "            return BestMotifs\n",
        "\n",
        "def Normalize(Probabilities):\n",
        "    # your code here\n",
        "    normProb = {}\n",
        "    for k,v in Probabilities.items():\n",
        "        normProb[k] = Probabilities[k]/sum(Probabilities.values())\n",
        "    return normProb\n",
        "\n",
        "def WeightedDie(Probabilities):\n",
        "    kmer = '' # output variable\n",
        "    # your code here\n",
        "    prob= Normalize(Probabilities)\n",
        "\n",
        "    sumProb=0\n",
        "    p= random.uniform(0,1)\n",
        "\n",
        "    for k,v in prob.items():\n",
        "        sumProb+= v\n",
        "        if p < sumProb:\n",
        "            kmer= k\n",
        "            break\n",
        "\n",
        "    return kmer\n",
        "\n",
        "def ProfileGeneratedString(Text, profile, k):\n",
        "    # your code here\n",
        "    Prob= {}\n",
        "    for i in range(len(Text)-k+1):\n",
        "        Prob[Text[i:i+k]]= MotifPr(profile, Text[i:i+k])\n",
        "\n",
        "    return WeightedDie(Prob)\n",
        "\n",
        "def GibbsSampler(Dna, k, t, N):\n",
        "    motifs= []\n",
        "    for seq in Dna:\n",
        "        i= random.randint(0, len(seq)-k)\n",
        "        motifs.append(seq[i:i+k])\n",
        "\n",
        "    BestMotifs = motifs # output variable\n",
        "\n",
        "    # your code here\n",
        "    for j in range(N):\n",
        "        l= random.randint(0, len(Dna)-1)\n",
        "        motifsNew= []\n",
        "\n",
        "        for i in range(len(motifs)):\n",
        "            if i != l:\n",
        "                motifsNew.append(motifs[i])\n",
        "\n",
        "        profile= ProfileWithPseudocounts(motifsNew)\n",
        "        motifs[l]= ProfileGeneratedString(Dna[l], profile, k)\n",
        "\n",
        "        if Score(motifs) < Score(BestMotifs):\n",
        "            BestMotifs= motifs\n",
        "    return BestMotifs"
      ],
      "metadata": {
        "id": "pDPTBBTX_QXN"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "DNA= ['ATGAGGTC', 'GCCCTAGA', 'AAATAGAT', 'TTGTGCTA']\n",
        "motifs= ['GTC', 'CCC', 'ATA', 'GCT']\n",
        "print(ProfileWithPseudocounts(motifs))\n",
        "Motifs(ProfileWithPseudocounts(motifs), DNA)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "CZBH7o2p87nD",
        "outputId": "ac55dc46-f657-4b08-ce61-4abf3f70b680"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "{'A': [0.25, 0.125, 0.25], 'C': [0.25, 0.375, 0.375], 'G': [0.375, 0.125, 0.125], 'T': [0.125, 0.375, 0.25]}\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "['GTC', 'GCC', 'ATA', 'GCT']"
            ]
          },
          "metadata": {},
          "execution_count": 6
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "kJnsj24sxLgX"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}